core/libraries/batch/JobHandlers/ExecuteBatchDeletion.php

```diff
@@ -147,16 +161,117 @@ public function continue_job(JobParameters $job_parameters, $batch_size = 50)
         return new JobStepResponse(
             $job_parameters,
             sprintf(
-                esc_html__('Deleted %d items.', 'event_espresso'),
+                // translators: 1: the quantity of items deleted.
+                esc_html__('Deleted %1$d items.', 'event_espresso'),
                 $units_processed
             )
         );
     }
 
+    /**
+     * @since $VID:$
+     * @param $reg_ids
+     * @param $batch_size
+     * @param JobParameters $job_parameters
+     * @return int
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws EE_Error
+     * @throws ReflectionException
+     */
+    protected function deleteRegistrationsAndTransactions($reg_ids, $batch_size, JobParameters $job_parameters)
+    {
+        $count_deleted = 0;
+        foreach ($reg_ids as $reg_id) {
+            $count_deleted += $this->deleteRegistrationAndTransaction($reg_id, $job_parameters);
+            if ($count_deleted >= $batch_size) {
+                return $count_deleted;
+            }
+        }
+        return $count_deleted;
+    }
+
+    /**
+     * @since $VID:$
+     * @param $reg_id
+     * @param JobParameters $job_parameters
+     * @return int
+     * @throws EE_Error
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws ReflectionException
+     */
+    protected function deleteRegistrationAndTransaction($reg_id, JobParameters $job_parameters)
+    {
+        // If the registration's transaction has no other valid registrations, delete it, and its dependent data too.
+        $TXN = EEM_Transaction::instance()->get_one(
+            [
+                [
+                    'Registration.REG_ID' => $reg_id
+                ]
+            ]
+        );
+        $count_deleted = 0;
+        if ($TXN instanceof EE_Transaction) {
+            $valid_regs = $TXN->count_related(
+                'Registration',
+                [
+                    [
+                        'REG_deleted' => false,
+                        'Event.status' => ['!=', EEM_Event::post_status_trashed]
+                    ]
+                ]
+            );
+            if (! $valid_regs) {
+                // delete the transaction
+                // and its dependent data
+                $transaction_as_root_of_tree = new ModelObjNode($TXN->ID(), $TXN->get_model());
+                $transaction_as_root_of_tree->visit(1000);
+                $ids_to_delete = $transaction_as_root_of_tree->getIds();
+                foreach ($ids_to_delete as $model_name => $ids) {
+                    // just grab the IDs the easy way, there's no Term_Relationship model related to transactions
+                    $model_to_delete = EE_Registry::instance()->load_model($model_name);
```

Comment:
confused by this comment... where are we grabbing the IDs other than from the foreach variable?

Reply:
Ya that doesn't make a lot of sense. I meant we can rely on the IDs actually just being simple int or string primary keys. They're not going to be those weird strings of key-value pairs we need to use for the `EEM_Term_Relationship` model (like we need to use for https://github.com/eventespresso/event-espresso-core/blob/CASC/base/core/libraries/batch/JobHandlers/ExecuteBatchDeletion.php#L113
If that makes sense I will update the comment (if not let's discuss some more.)